packages/core-transaction-pool/src/processor.ts

```diff
@@ -224,7 +227,15 @@ export class Processor implements TransactionPool.IProcessor {
     }
 
     private addTransactionsToPool(): void {
-        const { notAdded }: ITransactionsProcessed = this.pool.addTransactions(Array.from(this.accept.values()));
+        const acceptedTransactions = Array.from(this.accept.values());
+        // We need to decrement the nonce of the sender, since canBeApplied
+        // is called on all transactions before they are actually applied
+        // and the nonce lags behind.
+        for (const accepted of acceptedTransactions) {
+            this.walletManager.decrementNonce(accepted.data.senderPublicKey);
```

Comment:
This looks very bad - a hack that makes the code fragile in such a way that some future change could messup the nonce increment/decrements.

Is it not possible to have nonce increment in just one place when applying a transaction and decrement just in one place - when/if reverting a transaction?

Reply:
We already do increment/decrement when applying/reverting a transaction.

But without reworking a big part of the transaction pool processor, we need to increment/decrement nonces at this stage, too.

It's necessary because of how the pool works. We receive a payload of transactions and first perform some checks and call canBeApplied on each of them. At this stage the nonce does not change, but this violates the condition inside canBeApplied (that is if we receive multiple tx of the same sender in the same payload ).

Here's a super simplified version of how it works atm:
```
const acceptedTransactions = [];
for (const transaction of incomingTransactions) {
        ...
        a lot of checks
        canBeApplied(transaction)
        some more checks
        acceptedTransactions.push(transaction)
}
...
and more checks
...
addTransactionsToPool(acceptedTransactions) // wallet.apply is finally called here
```

How it needs to be to not need the hack:
```
for (const transaction of incomingTransactions) {
        ...
        some checks
        canBeApplied(transaction)
        more checks
        addTransactionToPool(transaction) // wallet.apply with each iteration
}
```

This way the nonce increases (if accepted) properly between each iteration and multiple transactions of the same sender don't cause issues anymore.

It's a bit out of scope for this PR.

Reply:
Ok, I agree - this is out of scope.